### DISCUSSION TITLE
Add native iOS SPM package in react native library

### DISCUSSION THREAD

**Discussion by yllfejziu** [callstack/react-native-builder-bob#505] [Q&A] [answered]
I created a react-native library and the example app which works fine.

However, I wish to add a native iOS library using **SPM** (since it's not on cocoapods).  The react-native library created by CLI created just 3 files:

`example.swift`
`example.mm`
`example-Bridging-Header.h`

Why isn't there a `.xcodeproj` automatically created when the library is scaffolded? How do I add one so I import the SPM package or is there any other way?

**Comment by atlj** [comment]
Hi, Swift Package Manager support for React Native is still in discussion. You can find the discussion [here](https://github.com/react-native-community/discussions-and-proposals/issues/587).

**Reply by yllfejziu** [reply]
Hey @atlj, thank you for the response.
I saw that cocoapods is usually the way to go with react native.
In my case, could you recommend any other way how to tackle this?
Thanks!

**Reply by Dwikavindra** [reply]
Hi I am also confused about this any solution for it ? @atlj

**Reply by yllfejziu** [reply]
Also for other people wondering...

Can you explain why isn't there a .xcodeproj, how to add one for people who need it?
Let's say you want to add a pod dependency to your react-native library, how do you do that?

Thanks in advance,
@atlj

**Reply by atlj** [reply]
To add a pod dependency, you can modify the .podspec file located at the root of your project. Those dependencies will be installed when a consumer installs your library.

**Reply by yllfejziu** [reply]
Yeah, got that, guess I didn't phrase the question appropriately. @atlj 

I meant if you need to use that dependency within the react-native library you're creating. For example:

`example.swift`
`example.mm`
`example-Bridging-Header.h`

As I mentioned on the discussion thread I created, the 3 files that are created, in `example.swift` how to import a native ios pod?

Don't you need an `.xcodeproj`?

**Reply by atlj** [reply]
@yllfejziu sorry if im not getting it right but by a native iOS pod, do you mean a Swift package? Specifically an SPM package? Because if so, currently React Native doesn't have support for SPM dependencies so even if you had the xcode project files, the added SPM dependencies won't be included on the library you build. At the current time, React Native doesn't install SPM dependencies automatically. Forgive my ignorance but I'd look for a way to create an xcframework from the mentioned spm package, and then I'd include it in the podspec file.
That being said, I'm still not 100% sure about SPM. If you think having the project file would help, you can try an older version of create-react-native-library (a version released around June 2023 should work) and let us know. If that helps to fix your problem, we can definitely bring the project file back. We deleted it previously because there was simply no use for it.

**Reply by yllfejziu** [reply]
I'll explain it from beginning in more depth. @atlj 

I created a native module (Kotlin & Swift) library using builder-bob tool. It generated the ios, android files, podspec, example app etc..

<img width="724" alt="Screenshot 2024-01-16 at 10 03 48 PM" src="https://github.com/callstack/react-native-builder-bob/assets/38266916/20ad3924-a026-4857-bdbe-06031fd1f034">

Now both iOS and android folders will be exposing some methods (initial template includes just a multiply function). 
For writing the native iOS and Android code we'll rely on some libraries and build on top of them, and then expose those functions.

<img width="855" alt="Screenshot 2024-01-16 at 10 07 14 PM" src="https://github.com/callstack/react-native-builder-bob/assets/38266916/cb9e5dfa-a8c6-4384-8230-5cdec3d8d212">

How do you import that pod here (not an SPM package, a cocoa pod)? 
If you declare it in the .podspec as a dependency, that'll be installed only when the consumer installs the library, but I want to use it within the library.

Please let me know if you have any questions since I've been trying to search for a solution online using CLI tools to scaffold libraries, but in some react native libraries I can see the `.xcodeproj` file, for example [this one](https://github.com/react-native-video/react-native-video).

Thanks!

p.s. for android it's easy I guess since you just declare it inside the `build.gradle` under `dependencies`.

**Reply by atlj** [reply]
The `.podspec` file is the iOS analog of build.gradle for your library. The dependencies under the `.podspec` file can be imported by your library without any problems. Once adding the pod to as a dependency to the `.podspec` file (Not to be confused with `Podfile`), you need to open `example/ios/<your_lib_name>.xcworkspace` using XCode. Then you'll be able to see your library under Pods/Development Pods. You can see the `CONTRIBUTING.md` file that's generated alongside your library for more information. Please feel free to ping me again if that didn't help.

**Reply by yllfejziu** [reply]
I don't need to import the library on the example app since this is the app that'll be using the library.
The code that I'm trying to build will use this dependency, build something on top of it, and provide it for the example app / any other app that'll use this native module.

To conclude:
I need to use the pod inside `ios/...`
not inside `example/ios/...`

And that's the reason I'm asking why there isn't an xcode project inside the ios folder.

Sorry for the misunderstandings, hopefully you get what I mean.

**Reply by Dwikavindra** [reply]
I guess in other words we are trying to ask why there is no `.xcodeproj` file inside the root ios folder while other packages for example https://github.com/react-native-video/react-native-video/tree/master/ios do have them. They have an `.xcodeproj` file in the root ios folder. As such we are able to open them in Xcode from the root ios folder not example/ios. The question then becomes how do you generate the file to begin with? @atlj

**Reply by atlj** [reply, marked as answer]
Folks sorry if I will sound like a broken record 😅 but the normal development workflow for libraries actually include opening the example/ios workspace in xcode and changing the code there. In the past, people were trying to open the project file located in the ios folder and they were getting issues since it's not the correct way of doing it. If you open the example app, go to the Pods target and Development Pods, you'll find your library code in the list there. If you edit the library code there, the changes will be reflected to your library (you can check the git diff).

**Reply by yllfejziu** [reply]
Oh, that's the piece I was missing 😅 
Thanks a lot, this is what I needed and it answers my question.

**Reply by atlj** [reply]
@yllfejziu that's awesome that it fixed the problem! I'd also highly suggest checking out the `CONTRIBUTING.md` file in the project you've generated. We've added a lot of useful stuff for library development there. Thanks a lot for the detailed explanation!

**Reply by yllfejziu** [reply]
Yeah, that's helpful, thanks.
Another question that is a little bit related to the development flow (considering the project structure as above), if you were to write unit tests in Swift, where would you put those? 
The example app already has a test target and then you can import your pod below, is that the correct flow (and not in the Pods target)? @atlj

**Reply by Andrew-Paystack** [reply]
Hey @atlj, 

Thanks a lot for answering this question. I'm curious why does the development flow for `Android` differ from `iOS`? 

In Android, we develop in the `android` folder while test using the `example/android` project.

**Reply by atlj** [reply]
@Andrew-Paystack I also do the development from `example/android`. Everything works well, and changes are reflected in fs.

**Reply by Andrew-Paystack** [reply]
Hey @atlj , 

Thanks for getting back to me. _ Please allow me to share my predicament in a bit more detail below and get more clarity._ 

I'm building a library that's implementing our Android and iOS SDKs to be distributed to the public.
When working on Android side, I built the Native implementation in the `android` folder. Then imported it in the `example/android` project. It worked and I've been able to test it on external projects.

In the `ios` folder, there are only 3 files generated, and I've not been able to get it to build on Xcode. Unlike android where there's a complete android project that's buildable and exportable. 

So my predicament becomes why would ios need to work from the examples folder while the android works from the root folder? 
Or is my approach wrong on some level that's yet to appear to me?

**Reply by satya164** [reply]
Both iOS and Android projects need to be developed using the example project as mentioned in your project's `CONTRIBUTING.md`. Any other workflow is unsupported.

**Reply by huy-lv** [reply]
Same question with @Andrew-Paystack , android folder contains complete android project, but there are only 3 files in ios folder, how can I open it in xcode? 
@atlj

**Reply by Kif11** [reply]
> Folks sorry if I will sound like a broken record 😅 but the normal development workflow for libraries actually include opening the example/ios workspace in xcode and changing the code there. In the past, people were trying to open the project file located in the ios folder and they were getting issues since it's not the correct way of doing it. If you open the example app, go to the Pods target and Development Pods, you'll find your library code in the list there. If you edit the library code there, the changes will be reflected to your library (you can check the git diff).

Thanks for clarifying! It would be great if this is included in the docs. At least in the FAQ section.

**Reply by Kif11** [reply]
I found that this is documented in generated `CONTRIBUTING.md` by `create-react-native-library` which I think is quite an obscure place. The logic is when I start my own module why would I read contribution guide in fact I didn't realize it exists until I start examining generated content closer.

**Reply by satya164** [reply]
> I found that this is documented in generated `CONTRIBUTING.md` by `create-react-native-library` which I think is quite an obscure place.

What would you say the appropriate place that you'd read before getting started?
